### PR TITLE
Amp caching module

### PR DESCRIPTION
**Pull request by prateekbh**
- Adds an amp caching module.
- Adds a test harness

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+const tmp = require('tmp');
```
Comment (kristoferbaxter) [inline index.js]: Worth considering, using `esm` with node so all source files are using the module specification.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+const tempDir = tmpobj.name;
+
+// Transpile TSC in a temp dir
+npmRun.sync(`tsc -p tsconfig.json --outDir ${tempDir}`);
+
+const filename = path.join(tempDir, 'index.js');
+const outfile = path.join(tempDir, 'build', 'index.js');
+
+// rollup the files in the tempdir
+npmRun.sync(`rollup ${filename} --config config/rollup.config.js --file ${outfile}`);
```
Comment (kristoferbaxter) [inline index.js]: If you're in JS land, you might as well use the rollup JS API. It's promise based and works well with `async/await`.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+// return the generated SW
+return fs.readFileSync(outfile, {
```
Comment (kristoferbaxter) [inline index.js]: Nit: Use promisify with `fs.readFile` and `await` the result.

File: package.json [hunk 4]
```diff
@@ -20,13 +19,39 @@
   },
   "homepage": "https://github.com/prateekbh/sw-generator#readme",
   "devDependencies": {
+    "@types/workbox-sw": "3.2.0",
+    "chai": "4.2.0",
+    "chromedriver": "^2.42.0",
+    "husky": "^1.1.2",
     "lint-staged": "7.3.0",
+    "mocha": "5.2.0",
+    "npm-run": "^5.0.1",
```
Comment (kristoferbaxter) [inline package.json]: There is a mix of hat versions and pinned versions, can we just use pinned versions and throw in renovate with the same configuration as other projects?

File: package.json [hunk 5]
```diff
@@ -4,9 +4,8 @@
   "description": "",
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1",
-    "prebuild": "tsc -p src/tsconfig.json --resolveJsonModule",
-    "build": "rollup --config config/rollup.config.js"
+    "test": "node ./test/index.js",
```
Comment (kristoferbaxter) [inline package.json]: Interesting moving all this to node instead of cmd. What's the rationale?

File: test/index.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
+const seleniumAssistant = require('selenium-assistant');
```
Comment (kristoferbaxter) [inline test/index.js]: Same feedback here as the other scripts, highly recommend using `esm` so all your code can be module driven.

File: test/index.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  await seleniumAssistant.downloadLocalBrowser('chrome', 'stable', expiration);
+
+  console.log('starting server...');
+  const serveDir = path.join(__dirname, '../');
+  server = serve(serveDir, {
+      port: 6881,
+  });
+
+  const browsers = seleniumAssistant.getLocalBrowsers();
+  browsers.forEach((browser) => {
```
Comment (kristoferbaxter) [inline test/index.js]: Nit: `browsers.forEach(browser => `

A singular argument doesn't require enclosure.

File: test/index.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    // Skip if the browser isn't stable.
+    if (browser.getReleaseName() !== 'stable') {
+      return;
+    }
+    // Print out the browsers name.
+    if(browser.getPrettyName() !== "Google Chrome Stable") {
+        return;
```
Comment (kristoferbaxter) [inline test/index.js]: Let's use `prettier` to avoid style conflicts when we paste code in from other repos.

File: test/index.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    console.log(`testing on ${browser.getPrettyName()}`);
+    browser.getSeleniumDriver()
+    .then(async (driver) => {
```
Comment (kristoferbaxter) [inline test/index.js]: `await` versus then?

File: test/index.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  // Run the tests.
+  mocha.run(function(failures){
+    process.exitCode = failures ? -1 : 0;  // exit with non-zero status if there were failures
+  }).on('end', function() {
+    seleniumAssistant.killWebDriver(driver);
+    server.stop();
+  });
+}
+
+initTest();
```
Comment (kristoferbaxter) [inline test/index.js]: If this file is going to just run the tests, do you need to create a named function and execute it? Or can the `initTest` method just be inlined and executed?

File: test/test-utils/sw-test-cleanup.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
+window.__testCleanup = () => {
+  const unregisterSW = () => {
```
Comment (kristoferbaxter) [inline test/test-utils/sw-test-cleanup.js]: This code would likely be more readable with `async/await`.

`(await window.caches.keys()).map(name => window.caches.delete(name))` etc.

File: test/test-utils/wait-for-sw-state.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
+window.__waitForSWState = (registration, desiredState) => {
+  return new Promise((resolve, reject) => {
+    let serviceWorker = registration.installing;
+
+    if (!serviceWorker) {
+      return reject(new Error('The service worker is not installing. ' + 'Is the test environment clean?'));
```
Comment (kristoferbaxter) [inline test/test-utils/wait-for-sw-state.js]: This string concatenation is for two static strings. Intentional?

File: test/test-utils/wait-for-sw-state.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    const stateListener = evt => {
+      if (evt.target.state === desiredState) {
+        serviceWorker.removeEventListener('statechange', stateListener);
+        return resolve();
```
Comment (kristoferbaxter) [inline test/test-utils/wait-for-sw-state.js]: Do you need to return the `resolve` method here?

File: test/amp-caching-sw.js [hunk 14]
```diff
@@ -0,0 +1,10 @@
+function ampAssetsCaching() {
```
Comment (kristoferbaxter) [inline test/amp-caching-sw.js]: This is the method in the `amp-caching` module right? Why duplicate into the test?

File: test/amp-caching/amp-caching-test.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
+describe('Amp caching', function() {
```
Comment (kristoferbaxter) [inline test/amp-caching/amp-caching-test.js]: Nit: `AMP Caching Module`

File: test/amp-caching/amp-caching-test.js [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+  let driver;
+
+  before(async () => {
+    driver = global.__AMPSW.driver;
```
Comment (kristoferbaxter) [inline test/amp-caching/amp-caching-test.js]: Is there a reason `driver` couldn't just be defined when its created?

`const driver - global.__AMPSW.driver;`

File: test/amp-caching/amp-caching-test.js [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+      await window.__testCleanup();
+      cb();
+    });
+  });
+
+  describe('Versioned JS', () => {
+    const versionedAmpRuntime = 'https://cdn.ampproject.org/rtv/001525381599226/v0.js';
+    //const versionedAmpExtension = 'https://cdn.ampproject.org/rtv/001525381599226/v0.js';
+
+    ['https://cdn.ampproject.org/rtv/001525381599226/v0.js', 'https://cdn.ampproject.org/rtv/001525381599226/v0.js'].forEach(() => {
```
Comment (kristoferbaxter) [inline test/amp-caching/amp-caching-test.js]: I'm confused, an array literal of the same two values as the const?

**Comment by prateekbh** [comment]
@kristoferbaxter my bad, you've caught me amidst all the setup process while making multiple choices simultaneously. i'll iron out all the nuances and review this together offline

File: test/amp-caching/amp-caching-test.js (hunk 17 above)
Comment (prateekbh) [reply test/amp-caching/amp-caching-test.js]: fixed to use all kinds of files that amp deploys

File: test/amp-caching/amp-caching-test.js (hunk 16 above)
Comment (prateekbh) [reply test/amp-caching/amp-caching-test.js]: fixed

File: test/amp-caching/amp-caching-test.js (hunk 15 above)
Comment (prateekbh) [reply test/amp-caching/amp-caching-test.js]: fixed

File: test/amp-caching-sw.js (hunk 14 above)
Comment (prateekbh) [reply test/amp-caching-sw.js]: removed this file to be auto generated by the test

File: test/index.js (hunk 9 above)
Comment (prateekbh) [reply test/index.js]: removed

File: test/index.js (hunk 8 above)
Comment (prateekbh) [reply test/index.js]: done

File: test/index.js (hunk 7 above)
Comment (prateekbh) [reply test/index.js]: i'll let prettier take care of this

File: test/index.js (hunk 6 above)
Comment (prateekbh) [reply test/index.js]: done

File: package.json (hunk 4 above)
Comment (prateekbh) [reply package.json]: fixed

File: index.js (hunk 3 above)
Comment (prateekbh) [reply index.js]: fixed

File: index.js (hunk 2 above)
Comment (prateekbh) [reply index.js]: done

File: index.js (hunk 1 above)
Comment (prateekbh) [reply index.js]: done

File: test/amp-caching/amp-caching-test.js [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+
+  before(async () => {
+    const generatedSW = await buildSW();
+    const promisifiedWriteFile = promisify(writeFile);
+    await promisifiedWriteFile(serviceWorkerPath, generatedSW);
+    await driver.get('http://localhost:6881/test/index.html');
+  });
+
+  after(async () => {
+    const promisifiedUnlink = promisify(unlink);
```
Comment (kristoferbaxter) [inline test/amp-caching/amp-caching-test.js]: I'd recommend just hoisting this to the top level, right after your `imports`.

```javascript
import * as fs from 'fs';
import { promisify } from 'util';

const unlink = promisify(fs.unlink);
```

File: src/index.ts [hunk 19]
```diff
@@ -1,6 +1,7 @@
-import {ampAssetsCaching} from './modules/amp-caching/index';
+import { ampAssetsCaching } from './modules/amp-caching/index';
+declare function importScripts(...urls: string[]): void;
```
[the remark this answers is not in the document]
Comment (prateekbh) [reply src/index.ts]: I'd rather follow your suggestion here, this is something i found on stackoverflow

File: src/modules/amp-caching/index.ts [hunk 20]
```diff
@@ -1,3 +1,26 @@
-export function ampAssetsCaching() {
+import _workbox from 'workbox-sw';
+declare const workbox: typeof _workbox;
 
+export function ampAssetsCaching() {
+  const versionedAssetsRE = /^https:\/\/cdn.ampproject.org\/rtv\/\d*\//;
```
[the remark this answers is not in the document]
Comment (prateekbh) [reply src/modules/amp-caching/index.ts]: Two things:
a. So far CC is not in scene
b. Does it matter if they get inlined?

File: test/amp-caching/amp-caching-test.js [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+import { join } from 'path';
+
+describe('AMP Caching Module', function() {
+  const driver = global.__AMPSW.driver;
+  const serviceWorkerPath = join('test', 'amp-caching-sw.js');
+  this.timeout(5000);
+
+  before(async () => {
+    const generatedSW = await buildSW();
+    const promisifiedWriteFile = promisify(writeFile);
```
[the remark this answers is not in the document]
Comment (prateekbh) [reply test/amp-caching/amp-caching-test.js]: done

File: test/amp-caching/amp-caching-test.js (hunk 18 above)
Comment (prateekbh) [reply test/amp-caching/amp-caching-test.js]: done

File: test/amp-caching/amp-caching-test.js [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+    await driver.executeAsyncScript(async cb => {
+      await window.__testCleanup();
+      const registration = await navigator.serviceWorker.register(
+        '/test/amp-caching-sw.js',
+      );
+      await window.__waitForSWState(registration, 'activated');
+      cb();
+    });
+    await driver.navigate().refresh();
+    await new Promise(resolve => setTimeout(resolve, 100));
```
Comment (kristoferbaxter) [inline test/amp-caching/amp-caching-test.js]: Both of these lines need a bit of explanation.

Is `100`ms just a lucky value? Can we instead listen to `document load` or similar on the page after refresh?

File: test/amp-caching/amp-caching-test.js [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+      await window.__testCleanup();
+      cb();
+    });
+  });
+
+  describe('Versioned JS', () => {
+    const ampRuntime = 'https://cdn.ampproject.org/rtv/001525381599226/v0.js';
+    const ampExtension =
+      'https://cdn.ampproject.org/rtv/001810022028350/v0/amp-mustache-0.1.js';
+    const filesToTest = [ampRuntime, ampRuntime];
```
Comment (kristoferbaxter) [inline test/amp-caching/amp-caching-test.js]: Likely meant to use `[ampRuntime, ampExtension]`.

File: test/amp-caching/amp-caching-test.js (hunk 22 above)
Comment (prateekbh) [reply test/amp-caching/amp-caching-test.js]: this was just a timeout thing, removed this

File: test/amp-caching/amp-caching-test.js (hunk 23 above)
Comment (prateekbh) [reply test/amp-caching/amp-caching-test.js]: Yep! fixed

File: src/modules/amp-caching/index.ts [hunk 24]
```diff
@@ -1,3 +1,28 @@
+import _workbox from 'workbox-sw';
+declare const workbox: typeof _workbox;
```
Comment (prateekbh) [inline src/modules/amp-caching/index.ts]: @kristoferbaxter this will go away in the following PR as the bundled workbox is working

**Pull request merged**